asteroid_filterbanks/pcen.py

```diff
@@ -0,0 +1,162 @@
+from torch import nn
+import torch
+from . import transforms
+from typing import Union
+
+try:
+    from typing import TypedDict
+except ImportError:  # Fallback for >= Python 3.7
+    from typing_extensions import TypedDict
+
+
+class ExponentialMovingAverage(nn.Module):
+    """
+    Computes the exponential moving average of an sequential input.
+
+    Influenced by leaf-audio's tensorflow implementation
+        https://github.com/google-research/leaf-audio/blob/7ead2f9fe65da14c693c566fe8259ccaaf14129d/leaf_audio/postprocessing.py#L27
+
+    See license here
+        https://github.com/google-research/leaf-audio/blob/master/LICENSE
+    """
+
+    def __init__(
+        self,
+        smooth: float = 0.04,
+        per_channel: bool = False,
+        n_channels: int = 2,
+        trainable: bool = False,
+    ):
+        super().__init__()
+        if per_channel:
+            self.weights = nn.Parameter(
+                torch.full((n_channels,), fill_value=smooth), requires_grad=trainable
+            )
+        else:
+            self.weights = nn.Parameter(
+                torch.full((1,), fill_value=smooth), requires_grad=trainable
+            )
+
+    def forward(self, mag_spec, initial_state):
```

Comment:
Could there be a default for initial state? 
Asking if it would make sense, not for an implementation yet.

Reply:
By the way, reading the code forward, it seems that this initial state only complicates the code, rather than simplifying it, is there a deeper reason for it being here that I don't see?

Reply:
This is a great point. 
From how I understand it, the `initial_state` can act like the hidden state in the RNN, i.e allow multiple sequential calls, instead of having to do one long one. 
I'll update the code to demostrate!

Reply:
I understand the idea, for real time implementation etc..  But it's not used here and makes things more complicated.
Maybe we could have a `cached_forward` (or smth like that) where the state is automatically saved, but I'm not sure it's worth it.

Reply:
It is really not too tricky to implement. 
[Here's a test that demonstrates it works.](https://github.com/asteroid-team/asteroid-filterbanks/pull/12/files?w=1#diff-362f8eb5ce493d9e043caaac8071b9a52d83af3a4dd28ebcd28f5dab4d44d09bR71)

Reply:
Indeed, it's not too tricky to implement, but I don't think the new way is better. 
Having it return a tuple makes it unintuitive. 

If we have a logic of stateful PCEN, it shouldn't be affecting the "classic" API that is unstateful.

Reply:
Honestly, I would simplify the code and remove all the statefulness. It complicates the code for no reason, and writing a stateful wrapper is really simple, for people that need it. 

Or, we could through a moderate, but larger effort, and enable both ways of working, stateful and not stateful, without negatively impacting the non-stateful interface.

Reply:
It might seem like too much review effort for a small piece of code, sorry if it seems in this way, but all details are important IMO :wink:

Reply:
@mpariente Selfishly, the real-time aspect is something very useful for my own projects, that I'd like to include it here. I don't think the idea of a tuple is too unintuitive in the PyTorch ecosystem. For example, [nn.RNN](https://pytorch.org/docs/stable/generated/torch.nn.GRU.html) returns a tuple of the output and hidden state. i.e.
```python
gru = nn.GRU(...)
output, hidden = gru(inputs)
```
The IIR smoothing filter is an important part of the PCEN, which inherently has state (like an RNN) so I feel like it makes sense.
But I can appreciate the ease of use of single input, single output forward. 

Can you give me some guidance on what would be the best API design-wise to have a different stateful vs stateless PCEN layer?

Here are some options I can think of....

Option 1
```python
stateful_pcen = StatefulPCEN()
energy, hidden = pcen(tf_rep)
...
pcen= PCEN()
energy = pcen(tf_rep)
```

Option 2
```python
stateful_pcen = PCEN(stateful=True)
energy, hidden = pcen(tf_rep)
...
pcen = PCEN(stateful=False) # The default
energy = pcen(tf_rep)
```

Option 3
```python
pcen = PCEN() # The default
energy = pcen(tf_rep, False) # The default, passing False returns no hidden
...
energy, hidden = pcen(tf_rep, None)
energy, hidden = pcen(tf_rep, hidden)
```
Or your own thoughts of course.

Reply:
The `StatefulPCEN` approach, with a `from_pcen` class method and a `reset_state` method would sound nice to me. 

This would keep common usage quite simple and enable easy conversion from one PCEN to the other. Would this sound ok to you?

Reply:
I don't think I fully follow @mpariente, like... 
```
pcen = PCEN(...)
energy = pcen(mag_spec)
....
stateful_pcen = StatefulPCEN.from_pcen(pcen)
energy, hidden = stateful_pcen(energy)
```
?

Reply:
I had in mind that `StatefulPCEN` would truly be stateful. 
```python
stateful_pcen = StatefulPCEN.from_pcen(pcen)
energy = stateful_pcen(energy)
state = stateful_pcen.current_state
# Pass on a new sentence, reset states
stateful_pcen.reset_states()
```

By what you describe is also good.

Reply:
@mpariente Changes made! I went with my purposed approach, as I think it plays nicer with exporting to things like onnxruntime.